components/layout_2020/taffy/construct.rs

```diff
@@ -0,0 +1,240 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::borrow::Cow;
+
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use style::values::specified::text::TextDecorationLine;
+
+use super::{TaffyContainer, TaffyItemBox, TaffyItemBoxInner};
+use crate::cell::ArcRefCell;
+use crate::context::LayoutContext;
+use crate::dom::{BoxSlot, LayoutBox, NodeExt};
+use crate::dom_traversal::{Contents, NodeAndStyleInfo, NonReplacedContents, TraversalHandler};
+use crate::flow::inline::construct::InlineFormattingContextBuilder;
+use crate::flow::{BlockContainer, BlockFormattingContext};
+use crate::formatting_contexts::{
+    IndependentFormattingContext, NonReplacedFormattingContext,
+    NonReplacedFormattingContextContents,
+};
+use crate::positioned::AbsolutelyPositionedBox;
+use crate::style_ext::DisplayGeneratingBox;
+
+impl TaffyContainer {
+    pub fn construct<'dom>(
+        context: &LayoutContext,
+        info: &NodeAndStyleInfo<impl NodeExt<'dom>>,
+        contents: NonReplacedContents,
+        propagated_text_decoration_line: TextDecorationLine,
+    ) -> Self {
+        let text_decoration_line =
+            propagated_text_decoration_line | info.style.clone_text_decoration_line();
+        let mut builder = TaffyContainerBuilder {
+            context,
+            info,
+            text_decoration_line,
+            contiguous_text_runs: Vec::new(),
+            jobs: Vec::new(),
+            has_text_runs: false,
+        };
+        contents.traverse(context, info, &mut builder);
+        builder.finish()
+    }
+}
+
+/// <https://drafts.csswg.org/css-flexbox/#flex-items>
+struct TaffyContainerBuilder<'a, 'dom, Node> {
+    context: &'a LayoutContext<'a>,
+    info: &'a NodeAndStyleInfo<Node>,
+    text_decoration_line: TextDecorationLine,
+    contiguous_text_runs: Vec<FlexTextRun<'dom, Node>>,
+    /// To be run in parallel with rayon in `finish`
+    jobs: Vec<FlexLevelJob<'dom, Node>>,
+    has_text_runs: bool,
+}
+
+enum FlexLevelJob<'dom, Node> {
+    /// Or pseudo-element
+    Element {
+        info: NodeAndStyleInfo<Node>,
+        display: DisplayGeneratingBox,
+        contents: Contents,
+        box_slot: BoxSlot<'dom>,
+    },
+    TextRuns(Vec<FlexTextRun<'dom, Node>>),
+}
+
```

Comment:
This change is implementing support for grid, but this is called `FlexLevelJob`. Perhaps a better name would be `GridLevelJob`?

Reply:
Ok, so I've made this change.

However, I've also (in a subsequent commit) gone ahead and extracted the part of the construction code which is shared between the flexbox and grid code (which is almost all of it) into its own file/abstraction (`layout_2020/modern_construct.rs`). This leaves only about 15 lines of code (excluding `use` statements)  in `flexbox/construct.rs` and `taffy/construct.rs`, so I've moved that code into the `mod.rs` files and remove the `construct.rs` modules.

This is in a separate commit, so we can roll this back if you don't want to land it as part of this PR. But IMO it's a nice improvement to not have this big chunk of code duplicated. And as far as I'm aware Flexbox and Grid are specced to handle children identically (in particular the wrapping of contiguous runs of text in an anonymous block node, which is the main thing this code does), so I would not anticipate this code needing to diverge in future.